### PR TITLE
server: add hlint hints to replace case analysis with combinators

### PR DESCRIPTION
**Pull request by abooij**
### Description

This PR implements 8 new hlint hints that enforce usage of our `onNothing` and `onLeft` combinators. Thus it further improves on #6126. Example hints provided by these new hlint hints:

```
src-lib/Hasura/RQL/Types/Common.hs:(292,20)-(294,57): Warning: Use onNothing
Found:
  case x of
    Just v -> return v
    Nothing -> fail "integer passed is out of bounds"
Perhaps:
  onNothing x (fail "integer passed is out of bounds")
```
```
src-lib/Hasura/RQL/DDL/RemoteRelationship/Validate.hs:(434,7)-(436,25): Warning: Use onLeft
Found:
  case eitherScalar of
    Left _ -> throwError $ InvalidGraphQLName $ toSQLTxt scalarType
    Right s -> pure s
Perhaps:
  onLeft
    eitherScalar
    (\ _c -> throwError $ InvalidGraphQLName $ toSQLTxt scalarType)
```

These new hints are triggered 23 times on the current master codebase.

This PR also removes some dead code, namely `coerceSet`, from `Hasura.Prelude`.

### Affected components

- [x] Server
- [x] Tests

### CODE REVIEW

File: server/src-lib/Hasura/Prelude.hs [hunk 1]
```diff
@@ -142,16 +139,6 @@ spanMaybeM f = go . toList
...
--- | Efficiently coerce a set from one type to another.
---
--- This has the same safety properties as 'Set.mapMonotonic', and is equivalent
--- to @Set.mapMonotonic coerce@ but is more efficient. This is safe to use when
--- both @a@ and @b@ have automatically derived @Ord@ instances.
---
--- https://stackoverflow.com/q/57963881/176841
-coerceSet :: Coercible a b=> Set.Set a -> Set.Set b
-coerceSet = unsafeCoerce
-
```
Comment (codingkarthik) [inline server/src-lib/Hasura/Prelude.hs:154]: Not sure why this is removed from here. Is this change relevant to this PR?

Comment (abooij) [reply server/src-lib/Hasura/Prelude.hs:154]: No, it's not relevant. Would you rather this change be merged separately?

Comment (codingkarthik) [reply server/src-lib/Hasura/Prelude.hs:154]: I guess it's just dead code removal. It's fine this way

**Review by codingkarthik — APPROVED** [review]
LGTM

File: server/src-lib/Hasura/Prelude.hs (hunk 1 above)
Comment (nicuveo) [reply server/src-lib/Hasura/Prelude.hs:154]: We should check that Pro doesn't use it.

Comment (abooij) [reply server/src-lib/Hasura/Prelude.hs:154]: @nicuveo According to some quick `grep`s, it's not used by pro.

Comment (nicuveo) [reply server/src-lib/Hasura/Prelude.hs:154]: Perfect, then. LGTM! :)

**Review by nicuveo — APPROVED** [review]

**Review by tirumaraiselvan — APPROVED** [review]
no cl required
